dart-impl/tasking/src/internal/dart_tasking_instrumentation.c

```diff
@@ -1,47 +1,282 @@
+    dart_tool_task_end_cb_t cb;
+    void *userdata;
+};
+struct dart_tool_task_cancel_cb {
+    dart_tool_task_cancel_cb_t cb;
+    void *userdata;
+};
+struct dart_tool_task_yield_leave_cb {
+    dart_tool_task_yield_leave_cb_t cb;
+    void *userdata;
+};
+struct dart_tool_task_yield_resume_cb {
+    dart_tool_task_yield_resume_cb_t cb;
+    void *userdata;
+};
+
+struct dart_tool_task_finalize_cb {
+    dart_tool_task_finalize_cb_t cb;
+    void *userdata;
+};
+
+struct dart_tool_task_add_to_queue_cb {
+    dart_tool_task_add_to_queue_cb_t cb;
+    void *userdata;
+};
+
+struct dart_tool_local_dep_cb {
+    dart_tool_local_dep_cb_t cb;
+    void *userdata;
+};
+
+struct dart_tool_remote_dep_cb {
+    dart_tool_remote_dep_cb_t cb;
+    void *userdata;
+};
 
+typedef struct dart_tool_task_create_cb dart_tool_task_create_cb;
+typedef struct dart_tool_task_begin_cb dart_tool_task_begin_cb;
+typedef struct dart_tool_task_end_cb dart_tool_task_end_cb;
+typedef struct dart_tool_task_cancel_cb dart_tool_task_cancel_cb;
+typedef struct dart_tool_task_yield_leave_cb dart_task_yield_leave_cb;
+typedef struct dart_tool_task_yield_resume_cb dart_task_yield_resume_cb;
+typedef struct dart_tool_task_finalize_cb dart_tool_task_finalize_cb;
+typedef struct dart_tool_task_add_to_queue_cb dart_tool_task_add_to_queue_cb;
+typedef struct dart_tool_local_dep_cb dart_tool_local_dep_cb;
+
+/* data structures to save the function pointer and user data from the callback */
+static struct dart_tool_task_create_cb dart_tool_task_create_cb_data;
+static struct dart_tool_task_begin_cb dart_tool_task_begin_cb_data;
+static struct dart_tool_task_end_cb dart_tool_task_end_cb_data;
+static struct dart_tool_task_cancel_cb dart_tool_task_cancel_cb_data;
+static struct dart_tool_task_yield_leave_cb dart_tool_task_yield_leave_cb_data;
+static struct dart_tool_task_yield_resume_cb dart_tool_task_yield_resume_cb_data;
+static struct dart_tool_task_finalize_cb dart_tool_task_finalize_cb_data;
+static struct dart_tool_task_add_to_queue_cb dart_tool_task_add_to_queue_cb_data;
+
+static struct dart_tool_local_dep_cb dart_tool_local_dep_cb_data;
+static struct dart_tool_remote_dep_cb dart_tool_remote_dep_cb_data;
+
+/* implementation of the register function of the callback */
+int dart_tool_register_task_create (dart_tool_task_create_cb_t cb, void *userdata_task_create) {
+   dart_tool_task_create_cb_data.cb = cb;
+   dart_tool_task_create_cb_data.userdata = userdata_task_create;
+   DART_LOG_TRACE("dart_tool_register_task_create was called\nPointer: %p and userdata %d",
+                  cb, *(int *) userdata_task_create);
+   return 0;
+} 
+
+int dart_tool_register_task_begin(dart_tool_task_begin_cb_t cb, void *userdata_task_begin) {
+    dart_tool_task_begin_cb_data.cb = cb;
+    dart_tool_task_begin_cb_data.userdata = userdata_task_begin;
+    DART_LOG_TRACE("dart_tool_register_task_begin was called\nPointer: %p and userdata %d",
+           cb, *(int *) userdata_task_begin);
```

Comment:
Don't dereference the user-provided argument, simply print the pointer. It might be `NULL` after all...